distributed/benchmarks/worker_state_machine.py

```diff
@@ -0,0 +1,151 @@
+            raise AssertionError(
+                f"Expected between {expect_min} and {expect_max} "
+                f"GatherDep instructions; got {actual}"
+            )
+
+
+class PopulateDataNeeded(_EnsureCommunicating):
+    """Create tasks in fetch state and push them into the data_needed heap"""
+
+    def setup(self, n_workers, n_tasks):
+        super().setup(n_workers, n_tasks)
+        self.ws.handle_stimulus(PauseEvent(stimulus_id="pause"))
+        self.ws.validate_state()
+        self.ws_pik = pickle.dumps(self.ws)
+
+    def time_unpickle(self, n_workers, n_tasks):
+        pickle.loads(self.ws_pik)
+
+    def time_populate_data_needed(self, n_workers, n_tasks):
+        ws = pickle.loads(self.ws_pik)
+        assert not ws.tasks
+        instructions = ws.handle_stimulus(self.acquire_replicas)
+        self.assert_instructions(instructions, 0, 0)
+        assert not ws.in_flight_workers
+
+
+class EnsureCommunicatingFromIdle(_EnsureCommunicating):
+    """Call _ensure_communicating while no workers are in flight"""
+
+    def setup(self, n_workers, n_tasks):
+        super().setup(n_workers, n_tasks)
+        self.ws.handle_stimulus(
+            PauseEvent(stimulus_id="pause"),
+            self.acquire_replicas,
+        )
+        self.ws.validate_state()
+        self.ws_pik = pickle.dumps(self.ws)
+
+    def time_unpickle(self, n_workers, n_tasks):
+        pickle.loads(self.ws_pik)
+
+    def time_from_idle(self, n_workers, n_tasks):
+        ws = pickle.loads(self.ws_pik)
+        assert not ws.in_flight_workers
+        instructions = ws.handle_stimulus(UnpauseEvent(stimulus_id="unpause"))
+        self.assert_instructions(instructions, 1, min(n_workers, 50))
+
+
+class EnsureCommunicatingNoop(_EnsureCommunicating):
+    """Call _ensure_communicating again while all workers are in flight"""
+
+    def setup(self, n_workers, n_tasks):
+        super().setup(n_workers, n_tasks)
+        self.ws.handle_stimulus(
+            self.acquire_replicas,
+            # FIXME this is only for compatibility with main
+            UnpauseEvent(stimulus_id="work-around-EnsureCommunicatingAfterTransitions"),
+            PauseEvent(stimulus_id="pause"),
+        )
+        self.ws.validate_state()
+
+    def time_noop(self, n_workers, n_tasks):
+        instructions = self.ws.handle_stimulus(UnpauseEvent(stimulus_id="unpause"))
+        self.assert_instructions(instructions, 0, 0)
+
+
+class EnsureCommunicatingOneWorker(_EnsureCommunicating):
+    """While all workers are in flight, one returns data, so another batch of tasks can
+    be fetched from the same worker
+    """
+
+    def setup(self, n_workers, n_tasks):
+        super().setup(n_workers, n_tasks)
+        instructions = self.ws.handle_stimulus(
+            self.acquire_replicas,
+            # FIXME this is only for compatibility with main
+            UnpauseEvent(stimulus_id="work-around-EnsureCommunicatingAfterTransitions"),
```

Comment:
FIXME do not merge
Merge https://github.com/dask/distributed/pull/6462 and then remove this